h3/src/client.rs

```diff
@@ -1,51 +1,58 @@
-use bytes::{Bytes, BytesMut};
-use futures::future;
-use http::{request, HeaderMap, Response};
 use std::{
     convert::TryFrom,
     marker::PhantomData,
     sync::{atomic::AtomicUsize, Arc},
     task::{Context, Poll, Waker},
 };
 
+use bytes::{Buf, Bytes, BytesMut};
+use futures::future;
+use http::{request, HeaderMap, Response};
+use tracing::{trace, warn};
+
 use crate::{
     connection::{self, ConnectionInner, ConnectionState, SharedStateRef},
     error::{Code, Error},
     frame::FrameStream,
     proto::{frame::Frame, headers::Header, varint::VarInt},
     qpack, quic, stream,
 };
-use tracing::{trace, warn};
 
-pub fn builder<C: quic::Connection<Bytes>>() -> Builder<C> {
+pub fn builder() -> Builder {
     Builder::new()
 }
 
-pub async fn new<C, O>(conn: C) -> Result<(Connection<C>, SendRequest<C::OpenStreams>), Error>
+pub async fn new<C, O>(conn: C) -> Result<(Connection<C, Bytes>, SendRequest<O, Bytes>), Error>
 where
     C: quic::Connection<Bytes, OpenStreams = O>,
     O: quic::OpenStreams<Bytes>,
 {
     Ok(Builder::new().build(conn).await?)
 }
 
-pub struct SendRequest<T: quic::OpenStreams<Bytes>> {
+pub struct SendRequest<T, B>
+where
+    T: quic::OpenStreams<B>,
+    B: Buf,
+{
     open: T,
     conn_state: SharedStateRef,
     max_field_section_size: u64, // maximum size for a header we receive
     // counts instances of SendRequest to close the connection when the last is dropped.
     sender_count: Arc<AtomicUsize>,
     conn_waker: Option<Waker>,
+    _buf: PhantomData<B>,
```

Comment:
Since this type never _holds_ a `B`, but instead just really _receives_ one and passes it on, you may want this to be `PhantomData<fn(B)>`. It's subtly different, and it alters how much auto traits (like `Send` and `Sync`) of `B` affect `SendRequest<T, B>`.

Reply:
Expert things! Yay ! If you can think of a link where I could learn more about this I'd be interested. Thanks !

Done in original commit.

Reply:
This comes from:

- [`std::marker::PhantomData`](https://doc.rust-lang.org/std/marker/struct.PhantomData.html#ownership-and-the-drop-check)
- [Nomicon](https://doc.rust-lang.org/nomicon/phantom-data.html)
- Experience and playing in the playground.

It'd be cool if there was a clearer resource that wasn't quite so deep in type theory...

Reply:
Thanks for sharing ! I should spend more time reading the Nomicon.